src/main/java/ch/njol/skript/lang/KeyProviderExpression.java

```diff
@@ -125,4 +161,111 @@ default boolean areKeysRecommended() {
+		//noinspection unchecked
+		Map.Entry<String, T>[] entries = new Map.Entry[values.length];
+		for (int i = 0; i < values.length; i++)
+			entries[i] = Map.entry(keys[i], values[i]);
+		return entries;
+	}
+
+	/**
+	 * Unzips an array of map entries into separate lists of keys and values.
+	 *
+	 * @param entries An array of map entries to unzip.
+	 * @param <T> The type of the values in the map entries.
+	 * @return An {@link Unzipped} object containing two lists: one for keys and one for values.
+	 */
+	static <T> Unzipped<T> unzip(@NotNull Map.Entry<String, T> @NotNull [] entries) {
+		List<String> keys = new ArrayList<>(entries.length);
+		List<T> values = new ArrayList<>(entries.length);
+		for (Map.Entry<String, T> entry : entries) {
+			keys.add(entry.getKey());
+			values.add(entry.getValue());
+		}
+		return new Unzipped<>(keys, values);
+	}
+
+	/**
+	 * Unzips an iterator of map entries into separate lists of keys and values.
+	 *
+	 * @param entries An iterator of map entries to unzip.
+	 * @param <T> The type of the values in the map entries.
+	 * @return An {@link Unzipped} object containing two lists: one for keys and one for values.
+	 */
+	static <T> Unzipped<T> unzip(Iterator<Map.Entry<String, T>> entries) {
+		List<String> keys = new ArrayList<>();
+		List<T> values = new ArrayList<>();
+		while (entries.hasNext()) {
+			Map.Entry<String, T> entry = entries.next();
+			keys.add(entry.getKey());
+			values.add(entry.getValue());
+		}
+		return new Unzipped<>(keys, values);
+	}
+
+	/**
+	 * Checks if the given expression can return keys.
+	 *
+	 * @param expression the expression to check
+	 * @return true if the expression can return keys, false otherwise
+	 * @see #canReturnKeys()
+	 */
+	static boolean canReturnKeys(Expression<?> expression) {
+		return expression instanceof KeyProviderExpression<?> provider && provider.canReturnKeys();
+	}
+
+	/**
+	 * Checks if the given expression can return keys and whether it is recommended to use them.
+	 *
+	 * @param expression the expression to check
+	 * @return true if the expression can return keys, and it is recommended to use them, false otherwise
+	 * @see #areKeysRecommended()
+	 * @see #canReturnKeys()
+	 */
+	static boolean areKeysRecommended(Expression<?> expression) {
+		return canReturnKeys(expression) && ((KeyProviderExpression<?>) expression).areKeysRecommended();
+	}
+
+	/**
+	 * A record that represents a pair of lists: one for keys and one for values.
+	 * This is used to store the result of unzipping map entries into separate lists.
+	 * <br>
+	 * Both lists are guaranteed to be of the same length, and each key corresponds to the value at the same index.
+	 *
+	 * @param <T> The type of the values in the list.
+	 * @param keys A list of keys extracted from the map entries.
+	 * @param values A list of values extracted from the map entries.
+	 * @see #unzip(Map.Entry[])
+	 * @see #unzip(Iterator)
+	 */
+	record Unzipped<T>(List<String> keys, List<T> values) {}
```

Comment:
i think unzipped is a fun name but i might prefer being a bit more explicit about it
eg UnzippedKeyValues or something more boring like that